gitclone/gitclone.go

```diff
@@ -65,71 +64,37 @@ func getMaxEnvLength() (int, error) {
 	return configs.EnvBytesLimitInKB * 1024, nil
 }
 
-func checkoutState(gitCmd git.Git, cfg Config) *step.Error {
-	checkoutArg := getCheckoutArg(cfg.Commit, cfg.Tag, cfg.Branch)
+func checkoutState(gitCmd git.Git, cfg Config) error {
+	checkoutMethod := selectCheckoutMethod(cfg)
+	fetchOpts := selectFetchOptions(checkoutMethod, cfg.CloneDepth, cfg.Tag != "")
 
-	isPR := cfg.PRRepositoryURL != "" || cfg.PRMergeBranch != "" || cfg.PRID != 0
-	if isPR {
-		if !cfg.ManualMerge || isPrivate(cfg.PRRepositoryURL) && isFork(cfg.RepositoryURL, cfg.PRRepositoryURL) {
-			if err := autoMerge(gitCmd, cfg.PRMergeBranch, cfg.BranchDest, cfg.BuildURL,
-				cfg.BuildAPIToken, cfg.CloneDepth, cfg.PRID); err != nil {
-				return newStepError(
-					"auto_merge_failed",
-					fmt.Errorf("merging PR (automatic) failed: %v", err),
-					"Merging pull request failed",
-				)
-			}
-		} else {
-			if err := manualMerge(gitCmd, cfg.RepositoryURL, cfg.PRRepositoryURL, cfg.Branch,
-				cfg.Commit, cfg.BranchDest); err != nil {
-				return newStepError(
-					"manual_merge_failed",
-					fmt.Errorf("merging PR (manual) failed: %v", err),
-					"Merging pull request failed",
-				)
-			}
-		}
-	} else if checkoutArg != "" {
-		if err := checkout(gitCmd, checkoutArg, cfg.Branch, cfg.CloneDepth, cfg.Tag != ""); err != nil {
-			return err
-		}
-		// Update branch: 'git fetch' followed by a 'git merge' is the same as 'git pull'.
-		if checkoutArg == cfg.Branch && cfg.Tag == "" && cfg.Commit == "" {
-			if err := runner.Run(gitCmd.Merge("origin/" + cfg.Branch)); err != nil {
-				return newStepError(
-					"update_branch_failed",
-					fmt.Errorf("updating branch (merge) failed %q: %v", cfg.Branch, err),
-					"Updating branch failed",
-				)
-			}
+	var patch string
+	if checkoutMethod == CheckoutPRDiffFileMethod {
+		var err error
+		if patch, err = getDiffFile(cfg.BuildURL, cfg.BuildAPIToken, cfg.PRID); err != nil {
```

Comment:
NOTE: instead of downloading the diff file here, might it would be cleaner to pass a diffFileDownloader to the `createCheckoutStrategy`

Reply:
Yes that was the plan, updated